src/Microsoft.Diagnostics.Monitoring.EventPipe/DiagnosticsEventPipeProcessor.cs

```diff
@@ -103,21 +95,30 @@ public async Task StopProcessing(CancellationToken token)
         {
             await _sessionStarted.Task;
 
-            EventPipeEventSource session = null;
-            Func<Task> stopFunc = null;
+            EventPipeEventSource eventSource = null;
+            EventPipeStreamProvider streamProvider = null;
+
             lock (_lock)
             {
-                session = _eventSource;
-                stopFunc = _stopFunc;
+                eventSource = _eventSource;
+
+                streamProvider = _streamProvider;
+                _streamProvider = null;
             }
-            if (session != null)
+
+            // Calling StopProcessing will not have the call to the Process method return until another
+            // event comes in. If no more events are going to come in, the thread that called the Process method
+            // will block until the EventPipeSession is disposed, which is owned by the EventPipeStreamProvider.
+            // Thus, dispose the stream provider in order to stop the session and event source in real time.
+            if (eventSource != null)
             {
-                //TODO This API is not sufficient to stop data flow.
-                session.StopProcessing();
+                eventSource.StopProcessing();
             }
-            if (stopFunc != null)
+
+            if (streamProvider != null)
             {
-                await stopFunc();
```

Comment:
So the only difference between EventPipeStreamProvider.Stop and Dispose is that the latter waits. I'm wondering if we should just remove the stop api altogether or have it behave the same way.